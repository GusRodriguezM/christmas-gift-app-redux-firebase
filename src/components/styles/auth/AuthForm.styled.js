import styled from 'styled-components';

const AuthForm = styled.form`
    width: 100%;
    height: 60%;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: space-evenly;
`;

export default AuthForm;