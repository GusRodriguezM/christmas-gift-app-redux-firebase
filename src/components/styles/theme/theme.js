const theme = {
    colors: {
        black: '#000000',
        loginButton: '#4285f4',
        disabledButton: '#999595',
        red: '#C30F16',
        white: '#FFFFFF',
        list: '#c8e4e4',
        navbar: '#e8ceac',
        boxShadow: 'rgba(9, 9, 9, 0.23)',
    }
}

export default theme;