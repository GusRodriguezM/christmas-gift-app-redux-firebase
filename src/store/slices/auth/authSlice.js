import { createSlice } from '@reduxjs/toolkit';

export const authSlice = createSlice({
  name: 'auth',
  initialState: {
    status: 'checking', //options: checking, non-authenticated, authenticated
    uid: null,
    email: null,
    userName: null,
    photoURL: null,
    errorMessage: null,
    logged: false
  },
  reducers: {
    login: (state, { payload }) => {
        state.status = 'authenticated';
        state.uid = payload.uid;
        state.email = payload.email;
        state.userName = payload.displayName;
        state.photoURL = payload.photoURL;
        state.errorMessage = payload.errorMessage;
    },
    logout: (state, { payload }) => {
        state.status = 'non-authenticated';
        state.uid = null;
        state.email = null;
        state.userName = null;
        state.photoURL = null;
        state.errorMessage = payload.errorMessage;
    },
    checkingUserCredentials: (state, action) => {
      state.status = 'checking';
    },
  },
});

export const { login, logout, checkingUserCredentials } = authSlice.actions;