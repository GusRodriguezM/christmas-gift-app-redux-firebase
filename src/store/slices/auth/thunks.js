import { signInWithGoogle } from "../../../firebase/providers";
import { checkingUserCredentials, login, logout } from "./"

export const checkingAuthentication = (email, password) => {
    return async( dispatch ) => {
        dispatch( checkingUserCredentials() );
    }
}

export const startGoogleSignIn = () => {
    return async( dispatch ) => {
        dispatch( checkingUserCredentials() );
        const result = await signInWithGoogle();
        
        if(!result.ok)
            return dispatch( logout( result.errorMessage ) );

        dispatch( login(result) );
    }
}