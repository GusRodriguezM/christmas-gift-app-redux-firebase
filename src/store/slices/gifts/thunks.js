import { collection, deleteDoc, doc, setDoc, writeBatch } from 'firebase/firestore/lite';
import { FirebaseDB } from '../../../firebase/config';
import { loadGifts } from '../../../helpers';
import { addGift, setSavingGift, setGifts, editGift, deleteGiftById, cleanList } from './';

export const startAddingNewGift = (newGift) => {
    return async(dispatch, getState) => {

        dispatch( setSavingGift() );
        
        const { uid } = getState().auth;

        const newDoc = doc( collection( FirebaseDB, `${uid}/list/gifts` ) );
        await setDoc( newDoc, newGift );

        newGift.id = newDoc.id;

        dispatch( addGift(newGift) );

    }
}

export const startLoadingGifts = () => {
    return async(dispatch, getState) => {

        const { uid } = getState().auth;
        if(!uid) throw new Error('The user id does not exist');

        const gifts = await loadGifts(uid);

        dispatch( setGifts(gifts) );
    }
}

export const startSavingGift = (giftToEdit) => {
    return async(dispatch, getState) => {
        dispatch( setSavingGift() );

        const { uid } = getState().auth;
        const { activeGift } = getState().gifts;

        const docRef = doc(FirebaseDB, `${uid}/list/gifts/${activeGift.id}`);
        await setDoc(docRef, giftToEdit, { merge: true });

        giftToEdit.id = activeGift.id;
        dispatch( editGift(giftToEdit) );
    }
}

export const startDeletingGift = (id) => {
    return async(dispatch, getState) => {

        const { uid } = getState().auth;

        const docRef = doc(FirebaseDB, `${uid}/list/gifts/${id}`);
        await deleteDoc(docRef);

        dispatch( deleteGiftById(id) );
        
    }
}

export const startDeletingGifts = () => {
    return async(dispatch, getState) => {
        
        //Getting the user id and the gifts array from the store
        const { uid } = getState().auth;
        const { gifts } = getState().gifts;

        //We use batch writes. Basically this instruction allows to execute a set of different operations
        const batch = writeBatch(FirebaseDB);

        //Loop over the array of gifts to delete. We send the reference to the doc
        gifts.forEach(gift => {
            batch.delete( doc(FirebaseDB, `${uid}/list/gifts/${gift.id}`) );
        });

        //Finally, we commit the batch
        await batch.commit();

        //We clean the array in the store
        dispatch( cleanList() );

    }
}