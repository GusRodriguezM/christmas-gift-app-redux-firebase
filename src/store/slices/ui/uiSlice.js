import { createSlice } from '@reduxjs/toolkit';

export const uiSlice = createSlice({
    name: 'modal',
    initialState: {
        msgError: ''
    },
    reducers: {
        setErrorMessage: (state, {payload}) => {
            state.msgError = payload;
        },
        removeErrorMessage: (state) => {
            state.msgError = null;
        },
    },
});

export const { setErrorMessage, removeErrorMessage } = uiSlice.actions;