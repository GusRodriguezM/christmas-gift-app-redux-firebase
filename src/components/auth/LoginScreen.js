import React, { useMemo, useEffect } from 'react';
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import validator from 'validator';

import { startGoogleSignIn, startLoginWithEmailPassword } from '../../store/slices/auth';
import { removeErrorMessage, setErrorMessage } from '../../store/slices/ui';
import { useForm } from '../../hooks';

import { Button } from '../styles/shared/Button.styled';
import Container from '../styles/auth/Container.styled';
import Input from '../styles/elements/Input.styled';
import Toast from '../styles/Toast/Toast';

export const LoginScreen = () => {

    const { status } = useSelector( state => state.auth );
    const { msgError } = useSelector( state => state.ui );
    const dispatch = useDispatch();

    const [formValues, handleInputChange] = useForm({
        email: '',
        password: ''
    });

    const { email, password } = formValues;

    const isAuthenticating = useMemo(() => status === 'checking', [status]);

    useEffect(() => {
        if(msgError.length > 0){
            Toast.fire({
                icon: 'error',
                title: `${msgError}`
            });
        }
    }, [msgError]);

    const isFormValid = () => {
        if(validator.isEmpty(email) && validator.isEmpty(password)){
            dispatch( setErrorMessage('Email and password are required') );
            return false;
        }else if(validator.isEmpty(email)){
            dispatch( setErrorMessage('The email is required') );
            return false;
        }else if(!validator.isEmail(email)){
            dispatch( setErrorMessage('The email is not valid') );
            return false;
        }else if(validator.isEmpty(password)){
            dispatch( setErrorMessage('The password is required') );
            return false;
        }

        dispatch( removeErrorMessage() );
        return true;
    }

    const handleLogin = (e) => {
        e.preventDefault();
        
        if(isFormValid())
            dispatch( startLoginWithEmailPassword({email, password}) );
    }

    const handleGoogleSignIn = () => {
        dispatch( startGoogleSignIn() );
    }

    return (
        <Container>

            <h1>Login</h1>

            <Container.AuthForm onSubmit={handleLogin}>

                <Input
                    id='name'
                    type='text'
                    placeholder='Email'
                    name='email'
                    autoComplete='off'
                    value={email}
                    onChange={handleInputChange}
                />

                <Input 
                    type='password'
                    placeholder='Password'
                    name='password'
                    autoComplete='off'
                    value={password}
                    onChange={handleInputChange}
                />

                <Button 
                    type='submit'
                    disabled={isAuthenticating}
                    inactive={isAuthenticating}
                >
                    <span>Login</span>
                    <i className="fa-solid fa-user-lock"></i>
                </Button>

                <Button 
                    onClick={handleGoogleSignIn}
                    disabled={isAuthenticating}
                    inactive={isAuthenticating}
                >
                    <span>Google Login</span>
                    <i className="fa-brands fa-google"></i>
                </Button>

            </Container.AuthForm>

            <Link
                aria-disabled={isAuthenticating}
                to='/auth/register'
            >
                Create a new account
            </Link>
        </Container>
    )
}