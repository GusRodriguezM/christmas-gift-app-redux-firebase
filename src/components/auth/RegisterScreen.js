import React, { useMemo, useEffect } from 'react';
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import validator from 'validator';

import { removeErrorMessage, setErrorMessage } from '../../store/slices/ui';
import { startCreatingUserWithEmailPassword } from '../../store/slices/auth';

import { useForm } from '../../hooks';

import Container from '../styles/auth/Container.styled';
import Input from '../styles/elements/Input.styled';
import { Button } from '../styles/shared/Button.styled';
import Toast from '../styles/Toast/Toast';

export const RegisterScreen = () => {

    const { msgError } = useSelector( state => state.ui );
    const { status } = useSelector( state => state.auth );
    const dispatch = useDispatch();

    const [formValues, handleInputChange] = useForm({
        name: '',
        email: '',
        password: '',
        password2: ''
    });

    const { name, email, password, password2 } = formValues;

    const isAuthenticating = useMemo(() => status === 'checking', [status]);

    useEffect(() => {
        if(msgError.length > 0){
            Toast.fire({
                icon: 'error',
                title: `${msgError}`
            });
        }
    }, [msgError]);
    

    const isFormValid = () => {
        if(validator.isEmpty(name) && validator.isEmpty(email) && validator.isEmpty(password) && validator.isEmpty(password2)){
            dispatch( setErrorMessage('Please fill all the required fields') );
            return false;
        }else if(validator.isEmpty(name)){
            dispatch( setErrorMessage('The name is required') );
            return false;
        }else if(validator.isEmpty(email)){
            dispatch( setErrorMessage('The email is required') );
            return false;
        }else if(!validator.isEmail(email)){
            dispatch( setErrorMessage('The email is not valid') );
            return false;
        }else if(password.length < 6 && password2.length < 6){
            dispatch( setErrorMessage('The password should be at least 6 characters') );
            return false;
        }else if(password !== password2){
            dispatch( setErrorMessage('The passwords should match') );
            return false;
        }

        dispatch( removeErrorMessage() );
        return true;
    }

    const handleRegister = (e) => {
        e.preventDefault();

        if(isFormValid()){
            dispatch( startCreatingUserWithEmailPassword({email, password, displayName: name}) );
        }
        
    }

    return (
        <Container>

            <h1>Sign in</h1>

            <Container.AuthForm 
                onSubmit={handleRegister}
                aria-label='register-form'>

                <Input
                    type='text'
                    placeholder='Name'
                    name='name'
                    autoComplete='off'
                    value={name}
                    onChange={handleInputChange}
                />

                <Input 
                    type='text'
                    placeholder='Email'
                    name='email'
                    autoComplete='off'
                    value={email}
                    onChange={handleInputChange}
                />

                <Input 
                    type='password'
                    placeholder='Password'
                    name='password'
                    autoComplete='off'
                    value={password}
                    onChange={handleInputChange}
                />

                <Input 
                    type='password'
                    placeholder='Password'
                    name='password2'
                    autoComplete='off'
                    value={password2}
                    onChange={handleInputChange}
                />

                <Button 
                    type='submit'
                    disabled={isAuthenticating}
                    inactive={isAuthenticating}
                >
                    <span>Register</span>
                    <i className="fa-solid fa-user-plus"></i>
                </Button>

            </Container.AuthForm>

            <Link
                aria-disabled={isAuthenticating}
                to='/auth/login'
            >
                Go to Login
            </Link>
            
        </Container>
    )
}