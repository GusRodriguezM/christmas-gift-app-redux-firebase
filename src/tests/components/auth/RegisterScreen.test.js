import { configureStore } from "@reduxjs/toolkit";
import { fireEvent, render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import { RegisterScreen } from "../../../components/auth/RegisterScreen";
import theme from "../../../components/styles/theme/theme";
import { authSlice } from "../../../store/slices/auth/authSlice";
import { uiSlice } from "../../../store/slices/ui/uiSlice";
import 'jest-styled-components';
import { ThemeProvider } from "styled-components";
import { nonAuthenticatedState } from "../../fixtures/authFixtures";
import { startCreatingUserWithEmailPassword } from "../../../store/slices/auth/thunks";

//Mocking the function of the thunk to begin with the creation of an user with email and password
const mockStartCreatingUserWithEmailPassword = jest.fn();
jest.mock('../../../store/slices/auth/thunks', () => ({
    startCreatingUserWithEmailPassword: ({email, password, displayName}) => {
        return () => mockStartCreatingUserWithEmailPassword({email, password, displayName})
    } 
}));

//Mocking the store with a fixture of the auth
const store = configureStore({
    reducer: {
        auth: authSlice.reducer,
        ui: uiSlice.reducer
    },
    preloadedState: {
        auth: nonAuthenticatedState
    }
});

describe('Tests in the <RegisterScreen />', () => {

    //Cleaning all the mocks
    beforeEach( () => jest.clearAllMocks() );

    test('Should render the component correctly and match with the snapshot', () => {

        //Rendering the component
        const { container } = render(
            <Provider store={ store } >
                <MemoryRouter>
                    <ThemeProvider theme={ theme } >
                        <RegisterScreen />
                    </ThemeProvider>
                </MemoryRouter>
            </Provider>
        );

        //Expecting that the container matches the snapshot
        expect( container ).toMatchSnapshot();

    });

    test('<RegisterScreen /> should have the components in the form', () => {

        //Rendering the component
        render(
            <Provider store={ store } >
                <MemoryRouter>
                    <ThemeProvider theme={ theme } >
                        <RegisterScreen />
                    </ThemeProvider>
                </MemoryRouter>
            </Provider>
        );

        //Expecting that the html tag h1 is in the DOM
        expect( screen.getByRole('heading', { level: 1, name: 'Sign in' }) ).toBeInTheDocument;
        
        //Getting the number of inputs and comparing with the real number in the component
        expect( screen.getAllByPlaceholderText('Name').length ).toBe(1);
        expect( screen.getAllByPlaceholderText('Email').length ).toBe(1);
        expect( screen.getAllByPlaceholderText('Password').length ).toBe(2);
        
        //Expecting that the button of the form is in the document
        expect( screen.getByRole('button', {name: 'Register'}) ).toBeInTheDocument;

        //Expecting that the form to have five children (the html elements tested above)
        const registerForm = screen.getByLabelText('register-form');
        expect(registerForm.children.length).toBe(5);

        //Getting the link and expecting to be in the document
        expect( screen.getByRole('link') ).toBeInTheDocument;

    });
    test('The register button should submit the form and call startCreatingUserWithEmailPassword with specific values', () => {

        //Making variables to send to the form
        const name = 'Bruce Wayne';
        const email = 'bruce.wayne@gotica.com';
        const password = '1mn0T84tw4N';
        const password2 = '1mn0T84tw4N';

        //Rendering the component
        render(
            <Provider store={ store } >
                <MemoryRouter>
                    <ThemeProvider theme={ theme } >
                        <RegisterScreen />
                    </ThemeProvider>
                </MemoryRouter>
            </Provider>
        );

        //Getting the input for the name and firing an event to put the corresponding value
        const nameField = screen.getByPlaceholderText('Name');
        fireEvent.change( nameField, { target: { name: 'name', value: name } } );

        //Getting the input for the email and firing an event to put the corresponding value
        const emailField = screen.getByPlaceholderText('Email');
        fireEvent.change( emailField, { target: { name: 'email', value: email } } )

        //Getting the inputs for the passwords and firing an event to put the corresponding value
        const passwordFields = screen.getAllByPlaceholderText('Password');
        fireEvent.change( passwordFields[0], { target: { name: 'password', value: password } } );
        fireEvent.change( passwordFields[1], { target: { name: 'password2', value: password2 } } );

        //Getting the form reference and firing an event to submit the data
        const registerForm = screen.getByLabelText('register-form');
        fireEvent.submit( registerForm );

        //Expecting that the thunk function have been called with the data we have put it in the form
        expect( mockStartCreatingUserWithEmailPassword ).toHaveBeenCalledWith( {email, password, displayName: name} );

    });

    test('Sweetalert should show if a field in the form is incorrect', () => {

        //Making variables to send to the form
        const name = 'Bruce Wayne';
        const email = 'bruce.wayne@gotica.com';
        const password = '84tw4N';
        const password2 = '1mn0T84tw4N';

        //Rendering the component
        render(
            <Provider store={ store } >
                <MemoryRouter>
                    <ThemeProvider theme={ theme } >
                        <RegisterScreen />
                    </ThemeProvider>
                </MemoryRouter>
            </Provider>
        );

        const nameField = screen.getByPlaceholderText('Name');
        fireEvent.change( nameField, { target: { name: 'name', value: name } } );

        const emailField = screen.getByPlaceholderText('Email');
        fireEvent.change( emailField, { target: { name: 'email', value: email } } )

        const passwordFields = screen.getAllByPlaceholderText('Password');
        fireEvent.change( passwordFields[0], { target: { name: 'password', value: password } } );
        fireEvent.change( passwordFields[1], { target: { name: 'password2', value: password2 } } );

        const registerForm = screen.getByLabelText('register-form');
        fireEvent.submit( registerForm );

        //Expecting that the error appears in the screen (with sweetalert) by selecting the html tag that contains the thrown error
        expect( screen.getAllByRole('heading', { level: 2, name: 'The passwords should match' }).length ).toBe(1);

    });

});