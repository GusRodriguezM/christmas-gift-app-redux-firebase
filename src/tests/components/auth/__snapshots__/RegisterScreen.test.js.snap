// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in the <RegisterScreen /> Should render the component correctly and match with the snapshot 1`] = `
.c1 {
  width: 100%;
  height: 80%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
}

.c0 {
  width: 450px;
  height: 450px;
  background: rgba( 211,226,236,0.5 );
  box-shadow: 0 8px 32px 0 rgba( 31,38,135,0.37 );
  -webkit-backdrop-filter: blur(4px);
  backdrop-filter: blur(4px);
  border-radius: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-top: 60px;
}

.c2 {
  margin: 10px 0;
  width: 95%;
}

.c3 {
  padding: 5px 5px;
  margin: 5px 0;
  text-align: center;
  color: #FFFFFF;
  border-radius: 5px;
  cursor: pointer;
  border: none;
  line-height: 1;
  width: 100px;
  background-color: #4285f4;
}

@media screen and (max-width:375px) {
  .c0 {
    width: 400px;
  }
}

@media screen and (max-width:425px) {
  .c3 span {
    display: none;
  }
}

@media screen and (min-width:426px) and (max-width:1440px) {
  .c3 i {
    display: none;
  }
}

@media screen and (max-width:375px) {
  .c3 {
    width: 50px;
  }
}

@media screen and (min-width:376px) and (max-width:1024px) {
  .c3 {
    width: 80px;
  }
}

<div>
  <div
    class="c0"
  >
    <h1>
      Sign in
    </h1>
    <form
      aria-label="register-form"
      class="c1"
    >
      <input
        autocomplete="off"
        class="c2"
        name="name"
        placeholder="Name"
        type="text"
        value=""
      />
      <input
        autocomplete="off"
        class="c2"
        name="email"
        placeholder="Email"
        type="text"
        value=""
      />
      <input
        autocomplete="off"
        class="c2"
        name="password"
        placeholder="Password"
        type="password"
        value=""
      />
      <input
        autocomplete="off"
        class="c2"
        name="password2"
        placeholder="Password"
        type="password"
        value=""
      />
      <button
        class="c3"
        type="submit"
      >
        <span>
          Register
        </span>
        <i
          class="fa-solid fa-user-plus"
        />
      </button>
    </form>
    <a
      aria-disabled="false"
      href="/auth/login"
    >
      Go to Login
    </a>
  </div>
</div>
`;
